### PR TITLE
[SPARK-38832][SQL][FOLLOWUP] Support propagate empty expression set for distinct key

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

- Improve `DistinctKeyVisitor` that support propagate empty set
- Small improvement for match alias

### Why are the changes needed?

Make distinct keys can be used to optimize more case, see comment https://github.com/apache/spark/pull/36117#discussion_r853755920

### Does this PR introduce _any_ user-facing change?

Improve performance

### How was this patch tested?

add test

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @cloud-fan @sigmod @wangyum

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala [hunk 1]
```diff
@@ -29,25 +29,21 @@ object DistinctKeyVisitor extends LogicalPlanVisitor[Set[ExpressionSet]] {
...
-              aliases
-                .collectFirst { case a: Alias if a.child.semanticEquals(expr) => a.toAttribute }
-                .getOrElse(expr)
-          }
-        }
-      }
+      val aliasedDistinctKeys = keys.map(_.map(_.transform {
+          case expr: Expression =>
+            aliases.get(expr.canonicalized).map(_.toAttribute).getOrElse(expr)
+        }))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala:42]:
```suggestion
      val aliasedDistinctKeys = keys.map(_.map(_.transform {
        case expr: Expression =>
          aliases.get(expr.canonicalized).map(_.toAttribute).getOrElse(expr)
      }))
```

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala:42]: nit: fix indentation

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala [hunk 2]
```diff
@@ -69,7 +65,9 @@ object DistinctKeyVisitor extends LogicalPlanVisitor[Set[ExpressionSet]] {
   override def default(p: LogicalPlan): Set[ExpressionSet] = Set.empty[ExpressionSet]
 
   override def visitAggregate(p: Aggregate): Set[ExpressionSet] = {
-    val groupingExps = ExpressionSet(p.groupingExpressions) // handle group by a, a
+    // handle group by a, a
+    // handle global aggregate
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/DistinctKeyVisitor.scala:69]:
```suggestion
    // handle group by a, a and global aggregate
```

**Review by cloud-fan — APPROVED** [review]

**Review by wangyum — COMMENTED** [review]
Could we filter out empty distinctKey for safe?
https://github.com/apache/spark/blob/21e48b7a2b88892835fdd9cd63662b07730a0103/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala#L442-L443

**Comment by cloud-fan** [comment]
@wangyum what's wrong with that code?

**Comment by wangyum** [comment]
For example:
```scala
Set[ExpressionSet](ExpressionSet()).exists(_.subsetOf(ExpressionSet(ae.aggregateFunction.children.filterNot(_.foldable))))
```
It is always true, it may have potential problems.

**Comment by ulysses-you** [comment]
if we get an unexpected empty ExpressionSet it should be a bug ..  shall we trust the framework ?

**Comment by cloud-fan** [comment]
I don't see why empty expression set is special. If the framework has bugs and produces incorrect distinct keys, we will have query correctness bugs.

**Comment by wangyum** [comment]
I'm ok if your do not think it is needed.

**Comment by cloud-fan** [comment]
My point is empty expression set is informative and we shouldn't skip it. It means the data is distinct by Nil, so it has at most 1 row and can override any other distict keys.

**Review by wangyum — APPROVED** [review]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by ulysses-you** [comment]
thank you @wangyum @cloud-fan

**Pull request closed without merging**